src/bin/lfortran.cpp

```
@@ -1188,9 +1188,9 @@ int link_executable(const std::vector<std::string> &infiles,
                 std::cout << "The command '" + cmd + "' failed." << std::endl;
                 return 10;
             }
-            if (outfile == "a.out") {
+            if (outfile == "a.out" && compiler_options.arg_o == "") {
                 err = system("a.out");
-                if (err != 0) return err;
+                if (err != 0) return 1;
```

What is the motivation to return 1 instead of "err"? Wouldn't it be better to simply propagate the error code?

---

> But `lpython examples/expr2.py` needs to return a non-zero exit code when the program fails to run. If it is not the case, we need to do it. [link](https://github.com/lcompilers/lpython/issues/1195#issuecomment-1276959879)

When we return `err`, then `lfortran` will return `0` on the failure. So I replaced it with 1.

---

Yes, we definitely need to return non zero exit code from LPython / LFortran.

But why does that happen? The if statement happens if `err` is non-zero, so how can it end up being zero? 

I wonder if there is some kind of wrap around happening: 256 -> 0 for 8 bit integer.

---

Yup, you are correct!
I tried to `return 256`, it printed `0`
But When I `return 255`, it prints `255`

---

I think it returns the exit code in the multiples of 256 which means 1 -> 256, 2 -> 512 and so on.

---

https://unix.stackexchange.com/questions/394639/why-do-high-exit-codes-on-linux-shells-256-not-work-as-expected

---

That works. To be sure, let's do this:
```c++
if (err != 0) {
  if (0 < err < 256) {
    return err;
  } else {
    return 1;
  }
}
```